fms/modules/tp.py

```diff
@@ -31,54 +25,205 @@ def setup_tp(self, rank: int, world_size: int) -> None:
+            Parameter that has had TP applied
+        tensor_value: torch.Tensor
+            tensor that needs sharding
+        is_bias: bool
+            if True is bias, else is weight
+        max_partition_sizes: List[int]
+            for each number in the list, if world_size is smaller than or equal to that number, the tensor will get
+            partitioned in worldsize parts, else if world size is larger than the number then you will get world size parts
+            replicated in worldsize / number
 
-    def import_weights(self, module: nn.Module):
-        for weight in self.colwise_param_names():
-            apply_colwise_tp(
-                getattr(self, weight),
-                getattr(module, weight),
-                self.world_size,
-                self.rank,
+            world_size = 4, max_partition_sizes = [8], tensor = [0 1 2 3 4 5 6 7]
+            [0 1] [2 3] [4 5] [6 7]
+
+            world_size = 8, max_partition_sizes = [4], tensor = [0 1 2 3 4 5 6 7]
+            [0 1] [0 1] [2 3] [2 3] [4 5] [4 5] [6 7] [6 7]
+
+            If there are multiple numbers in the max_partition_sizes list, then the param gets filled with non-contiguous
+            slices of the tensor_value. This is useful for fused weight cases (qkv, mlp, moe, etc.)
+
+            world_size = 4, max_partition_sizes = [4, 4], tensor = [0 1 2 3 4 5 6 7]
+            [0 4] [1 5] [2 6] [3 7]
+
+            world_size = 4, max_partition_sizes = [4, 1], tensor = [0 1 2 3 4 5 6 7 8 9]
+            [0 1 8 9] [2 3 8 9] [4 5 8 9] [6 7 8 9]
+        """
+        # Divide the weight matrix along the first dimension.
+        output_size_per_partition = param.shape[0] // (
+            sum(max_partition_sizes) // min(max_partition_sizes)
+        )
+        if not is_bias:
+            tp_slices = self.__get_tp_slices(
+                tensor_value.shape[0], output_size_per_partition, max_partition_sizes
             )
-        for weight in self.rowwise_param_names():
-            apply_rowwise_tp(
-                getattr(self, weight),
-                getattr(module, weight),
-                self.world_size,
-                self.rank,
+            tensors_to_cat = [tensor_value[tp_slice, :] for tp_slice in tp_slices]
+            tensor = torch.cat(tensors_to_cat, dim=0)
+        else:
+            tp_slices = self.__get_tp_slices(
+                tensor_value.shape[0], output_size_per_partition, max_partition_sizes
             )
-        for weight in self.embedding_param_names():
-            apply_embedding_tp(
-                getattr(self, weight),
-                getattr(module, weight),
-                self.world_size,
-                self.rank,
+            tensors_to_cat = [tensor_value[tp_slice] for tp_slice in tp_slices]
+            tensor = torch.cat(tensors_to_cat, dim=0)
+        param.copy_(tensor, non_blocking=True)
+
+    def _get_sd_weight(
+        self, state_dict: Dict[str, torch.Tensor], conditions: List[str]
```

Comment:
should have a comment (or different name?) for `conditions` ... looks like it's a set of simple string match patterns not a representation of logic.